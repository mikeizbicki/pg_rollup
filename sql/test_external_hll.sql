SET client_min_messages TO WARNING;
create or replace language plpython3u;
create extension if not exists hll;
drop extension pgrollup;
create extension if not exists pgrollup;

create table test (
    id serial primary key,
    name text,
    num int
);

insert into test (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);

CREATE MATERIALIZED VIEW test_rollup1 AS (
    SELECT
        name,
        round(hll_cardinality(hll_add_agg(hll_hash_text(name)))) AS distinct_name,
        round(hll_cardinality(hll_add_agg(hll_hash_integer(num)))) AS distinct_num
    FROM test
    GROUP BY name
);

CREATE MATERIALIZED VIEW test_rollup2 AS (
    SELECT
        name,
        hll_count(name) AS distinct_name,
        hll_count(num)  AS distinct_num
    FROM test
    GROUP BY name
);

select assert_rollup('test_rollup1');
select assert_rollup('test_rollup2');

insert into test (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);


select assert_rollup('test_rollup1');
select assert_rollup('test_rollup2');

select * from test_rollup1;
select * from test_rollup2;

drop table test cascade;
