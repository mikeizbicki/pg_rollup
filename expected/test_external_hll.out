SET client_min_messages TO WARNING;
create table test (
    id serial primary key,
    name text,
    num int
);
insert into test (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);
CREATE MATERIALIZED VIEW test_rollup1 AS (
    SELECT
        name,
        round(hll_cardinality(hll_add_agg(hll_hash_text(name)))) AS distinct_name,
        round(hll_cardinality(hll_add_agg(hll_hash_integer(num)))) AS distinct_num
    FROM test
    GROUP BY name
);
CREATE MATERIALIZED VIEW test_rollup2 AS (
    SELECT
        name,
        hll_count(name) AS distinct_name,
        hll_count(num)  AS distinct_num
    FROM test
    GROUP BY name
);
select assert_rollup('test_rollup1');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('test_rollup2');
 assert_rollup 
---------------
 
(1 row)

insert into test (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);
select assert_rollup('test_rollup1');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('test_rollup2');
 assert_rollup 
---------------
 
(1 row)

select * from test_rollup1;
 distinct_name | distinct_num | test.name 
---------------+--------------+-----------
             0 |            3 | 
             1 |            5 | alice
             1 |            1 | bill
             1 |            2 | charlie
             1 |            1 | dave
             1 |            1 | elliot
(6 rows)

select * from test_rollup2;
 distinct_name | distinct_num | test.name 
---------------+--------------+-----------
             0 |            3 | 
             1 |            5 | alice
             1 |            1 | bill
             1 |            2 | charlie
             1 |            1 | dave
             1 |            1 | elliot
(6 rows)

drop table test cascade;
